ui/src/dashboards/components/GraphOptionsCustomizableField.tsx

```diff
@@ -1,70 +1,186 @@
+const FieldType = 'field'
 
 interface Field {
   internalName: string
   displayName: string
   visible: boolean
+  order?: number
+  index?: number
 }
 
-interface Props {
+interface FieldSourceProps {
   internalName: string
   displayName: string
   visible: boolean
-  onFieldUpdate: (field: Field) => void
+  index: number
+  id: string
+  key: string
+  isDragging?: boolean
+  onFieldUpdate?: (field: Field) => void
+  connectDragSource?: any
+  connectDropTarget?: any
+  connectDragPreview?: any
+  moveField: (dragIndex: number, hoverIndex: number) => void
 }
 
-class GraphOptionsCustomizableField extends PureComponent<Props, {}> {
+const fieldSource: DragSourceSpec<FieldSourceProps> = {
+  beginDrag(props) {
+    return {
+      id: props.id,
+      index: props.index,
+    }
+  },
+}
+
+const fieldTarget = {
+  hover(props, monitor, component) {
+    const dragIndex = monitor.getItem().index
+    const hoverIndex = props.index
+
+    // Don't replace items with themselves
+    if (dragIndex === hoverIndex) {
+      return
+    }
+
+    // Determine rectangle on screen
+    const hoverBoundingRect = findDOMNode(component).getBoundingClientRect()
+
+    // Get vertical middle
+    const hoverMiddleY = (hoverBoundingRect.bottom - hoverBoundingRect.top) / 2
+
+    // Determine mouse position
+    const clientOffset = monitor.getClientOffset()
+
+    // Get pixels to the top
+    const hoverClientY = clientOffset.y - hoverBoundingRect.top
+
+    // Only perform the move when the mouse has crossed half of the items height
+    // When dragging downwards, only move when the cursor is below 50%
+    // When dragging upwards, only move when the cursor is above 50%
+
+    // Dragging downwards
+    if (dragIndex < hoverIndex && hoverClientY < hoverMiddleY) {
+      return
+    }
+
+    // Dragging upwards
+    if (dragIndex > hoverIndex && hoverClientY > hoverMiddleY) {
+      return
+    }
+    // Time to actually perform the action
+    props.moveField(dragIndex, hoverIndex)
+
+    // Note: we're mutating the monitor item here!
+    // Generally it's better to avoid mutations,
+    // but it's good here for the sake of performance
+    // to avoid expensive index searches.
+    monitor.getItem().index = hoverIndex
+  },
+}
+
+function MyDropTarget(dropv1, dropv2, dropfunc1) {
+  return target => DropTarget(dropv1, dropv2, dropfunc1)(target) as any
+}
+
+function MyDragSource(dragv1, dragv2, dragfunc1) {
+  return target => DragSource(dragv1, dragv2, dragfunc1)(target) as any
+}
+
+@MyDropTarget(FieldType, fieldTarget, (connect: DropTargetConnector) => ({
+  connectDropTarget: connect.dropTarget(),
+}))
+@MyDragSource(
+  FieldType,
+  fieldSource,
+  (connect: DragSourceConnector, monitor: DragSourceMonitor) => ({
+    connectDragSource: connect.dragSource(),
+    connectDragPreview: connect.dragPreview(),
+    isDragging: monitor.isDragging(),
+  })
+)
+export default class GraphOptionsCustomizableField extends PureComponent<
+  FieldSourceProps
```

Comment:
could these props be named something like GraphOptionsCustomizableField so its more intuitive that they are the props for this component